backend/ethereum/subscription/resistanteventsub_test.go

```diff
@@ -0,0 +1,229 @@
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+	numTx := 5
+	finality := int(rng.Int31n(maxFinality) + 1)
+
+	// Send some past tx.
+	for i := 0; i < numTx; i++ {
+		s.IncAllowance(ctx)
+	}
+
+	// Create a sub to go `numTx+finality` into the past.
+	sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, uint64(numTx), uint64(finality))
+	require.NoError(err)
+	defer sub.Close()
+
+	// Finalize the past events by mining blocks.
+	for i := 0; i < finality-1; i++ {
+		s.SB.Commit()
+	}
+	// Send some future tx.
+	for i := 0; i < numTx; i++ {
+		s.IncAllowance(ctx)
+	}
+	// Finalize the future events by mining blocks.
+	for i := 0; i < finality-1; i++ {
+		s.SB.Commit()
+	}
+
+	// Receive exactly `numTx` past events.
+	{
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+		sink := make(chan *subscription.Event, numTx+1)
+		err := sub.ReadPast(ctx, sink) // read only past events
+		require.NoError(err)
+
+		for i := 0; i < numTx; i++ {
+			require.NotNil(<-sink)
+		}
+
+		select {
+		case event := <-sink:
+			require.Nil(event)
+		default:
+		}
+	}
+}
+
+// TestResistantEventSub_ReorgConfirm tests that a TX is confirmed exactly
+// after `finalityDepth` blocks even when reorgs occur.
+func TestResistantEventSub_ReorgConfirm(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	rng := pkgtest.Prng(t)
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	finality := rng.Int31n(maxFinality-2) + 2
+	sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+	require.NoError(err)
+	defer sub.Close()
+
+	// Allow the reorg to go up to `maxFinality` blocks into the past.
+	for i := 0; i < maxFinality; i++ {
+		s.SB.Commit()
+	}
+
+	// Send and Confirm the TX
+	s.IncAllowance(ctx)
+	// Reorg until the block hight hits `finality` blocks.
+	for h := int64(0); h < int64(finality-1); {
+		d := rng.Int63n(maxFinality/2-1) + 1
+		l := rng.Int63n(maxFinality/2-1) + 1
+		NoEvent(require, sub)
+		log.Debugf("[h=%d] Reorg with depth: %d, length: %d", h, d, l)
+		s.SB.Reorg(ctx, uint64(d), func(txs []types.Transactions) []types.Transactions {
+			return append(txs, make([]types.Transactions, int(d+l)-len(txs))...)
+		})
+		h += l
+	}
+	// Verify that the event arrived.
+	OneEvent(require, sub)
+}
+
+// TestResistantEventSub_ReorgRemove tests that a TX is never confirmed if it
+// was removed by a reorg.
+func TestResistantEventSub_ReorgRemove(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	rng := pkgtest.Prng(t)
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	finality := rng.Int31n(maxFinality-2) + 2
+	sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+	require.NoError(err)
+	defer sub.Close()
+
+	// Send and Confirm the TX
+	s.ConfirmTx(s.IncAllowance(ctx), true)
```

Comment:
I have trouble following. Is it correct, that `ConfirmTx` waits until `TxFinalityDepth = 6` blocks until it returns? How is that the case here, if we do not progress the chain?

Update:
OK, checked in more detail. Realized that you set `TxFinalityDepth = 1` in the init function for this test. It doesn't seem like a good idea that we specify `TxFinalityDepth` as a global variable. It's hard to follow which value it currently has and there is also no protection against changing it on-the-fly, which may result in undefined behavior. Anyways, this is out of scope here. I will create an issue for that.

Reply:
Do we need to call `ConfirmTx` here at all?

Reply:
Right, i just added that yesterday for debugging, removed.